runtime/parachains/src/scheduler.rs

```diff
@@ -13,3 +13,514 @@
+#[derive(Encode, Decode, Default, Clone, Copy)]
+pub struct GroupIndex(u32);
+
+/// A claim on authoring the next block for a given parathread.
+#[derive(Clone, Encode, Decode, Default)]
+pub struct ParathreadClaim(pub ParaId, pub CollatorId);
+
+/// An entry tracking a claim to ensure it does not pass the maximum number of retries.
+#[derive(Clone, Encode, Decode, Default)]
+pub struct ParathreadEntry {
+	claim: ParathreadClaim,
+	retries: u32,
+}
+
+/// A queued parathread entry, pre-assigned to a core.
+#[derive(Encode, Decode, Default)]
+pub struct QueuedParathread {
+	claim: ParathreadEntry,
+	core_offset: u32,
+}
+
+/// The queue of all parathread claims.
+#[derive(Encode, Decode, Default)]
+pub struct ParathreadClaimQueue {
+	queue: Vec<QueuedParathread>,
+	// this value is between 0 and config.parathread_cores
+	next_core_offset: u32,
+}
+
+impl ParathreadClaimQueue {
+	// Queue a parathread entry to be processed.
+	//
+	// Provide the entry and the number of parathread cores, which must be greater than 0.
+	fn queue_entry(&mut self, entry: ParathreadEntry, n_parathread_cores: u32) {
+		let core_offset = self.next_core_offset;
+		self.next_core_offset = (self.next_core_offset + 1) % n_parathread_cores;
+
+		self.queue.push(QueuedParathread {
+			claim: entry,
+			core_offset,
+		})
+	}
+
+	// Take next queued entry with given core offset, if any.
+	fn take_next_on_core(&mut self, core_offset: u32) -> Option<ParathreadEntry> {
+		let pos = self.queue.iter().position(|queued| queued.core_offset == core_offset);
+		pos.map(|i| self.queue.remove(i).claim)
+	}
+}
+
+/// What is occupying a specific availability core.
+#[derive(Clone, Encode, Decode)]
+pub(crate) enum CoreOccupied {
+	Parathread(ParathreadEntry),
+	Parachain,
+}
+
+/// The assignment type.
+#[derive(Clone, Encode, Decode)]
+pub enum AssignmentKind {
+	Parachain,
+	Parathread(CollatorId, u32),
+}
+
+/// How a free core is scheduled to be assigned.
+#[derive(Encode, Decode)]
+pub struct CoreAssignment {
+	/// The core that is assigned.
+	pub core: CoreIndex,
+	/// The unique ID of the para that is assigned to the core.
+	pub para_id: ParaId,
+	/// The kind of the assigment.
+	pub kind: AssignmentKind,
+	/// The index of the validator group assigned to the core.
+	pub group_idx: GroupIndex,
+}
+
+impl CoreAssignment {
+	/// Get the ID of a collator who is required to collate this block.
+	pub(crate) fn required_collator(&self) -> Option<&CollatorId> {
+		match self.kind {
+			AssignmentKind::Parachain => None,
+			AssignmentKind::Parathread(ref id, _) => Some(id),
+		}
+	}
+}
+
+/// Reasons a core might be freed
+pub enum FreedReason {
+	/// The core's work concluded and the parablock assigned to it is considered available.
+	Concluded,
+	/// The core's work timed out.
+	TimedOut,
+}
+
+pub trait Trait: system::Trait + configuration::Trait + paras::Trait { }
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Scheduler {
```

Comment:
Currently clashes with the Substrate `Scheduler`: https://github.com/paritytech/substrate/blob/master/frame/scheduler/src/lib.rs#L100

Reply:
Or will the storage keys actually clash? :(